Источник:

https://data-in.ru/data-catalog/datasets/187/#dataset-codebook

In [1]:
import pandas as pd
import numpy as np

# from google.colab import drive
# drive.mount('/content/drive')

Для эксперимента возьмём датасет "Крупные города России: объединенные данные по основным социально-экономическим показателям за 1985-2019 гг." инфраструктуры научно-исследовательских данных (https://data-in.ru/data-catalog/datasets/187/#dataset-codebook). Использовать будем только столбцы 'region' (субъект РФ), 'municipality' (муниципальное образование), 'year' (год), 'birth' (число родившихся на 1000 человек населения), 'wage' (cреднемесячная номинальная начисленная заработная плата, руб.). Сразу оговорюсь, что Москва и Санкт-Петербург являются отдельными субъектами Российской Федерации и в этом датафрейме отсутствуют.

In [39]:
df1 = pd.read_csv('Krupnie_goroda-RF_1985-2019_187_09.12.21/data.csv', sep=';')
print(df.columns)
df = df1[['region', 'municipality', 'year', 'birth', 'wage']]
df.sample(20)

Index(['region', 'municipality', 'year', 'birth', 'wage'], dtype='object')


region                              municipality  year  \
3890     Республика Татарстан                          Набережные Челны  2017   
3388  Республика Башкортостан                               Октябрьский  2018   
2862        Тюменская область                                    Тюмень  2006   
189         Красноярский край                                Красноярск  2017   
1015        Иркутская область                           Ангарский район  2015   
2645     Свердловская область              Асбестовский городской округ  2003   
681      Владимирская область                            Город Владимир  2003   
798     Волгоградская область                             Город Камышин  2013   
2385       Ростовская область                             Город Батайск  2011   
2371       Ростовская область                      Город Ростов-на-Дону  2015   
677      Владимирская область                            Город Владимир  1999   
250           Приморский край                           Владивостокский  2005   
3211       Забайкальский край                                      Чита  2009   
3164      Челябинская область                            Магнитогорский  2009   
2403       Ростовская область                          Город Волгодонск  2009   
1246      Кемеровская область                                 Беловский  2002   
1200          Камчатский край  Петропавловск-Камчатский городской округ  2003   
2669     Свердловская область                        город Нижний Тагил  2003   
630      Белгородская область           Старооскольский городской округ  2006   
1158        Калужская область                              город Калуга  2008   

      birth     wage  
3890   13.2  33841.5  
3388   12.0  35612.2  
2862   12.6  16702.0  
189    12.8  44384.3  
1015   14.3  38856.3  
2645   10.4   4704.0  
681     9.0   4807.0  
798     9.6  18117.8  
2385   14.3  19783.2  
2371   12.5  35394.7  
677     6.5   1298.6  
250     9.2  10267.0  
3211   17.2  21807.2  
3164   13.4  19286.6  
2403   10.4  19008.8  
1246   10.4   4827.0  
1200   10.8  11685.0  
2669   10.0   5796.0  
630    10.0  11097.3  
1158    9.9  16012.7

In [40]:
print(df.shape)
df.columns


(4109, 5)


Index(['region', 'municipality', 'year', 'birth', 'wage'], dtype='object')

Минимальные данные можно получить использовав метод .describe(include = 'all'). Мы видим что у нас 4109 строки, по 81 региону и 202 городам. Средняя рождаемость на 1000 человек 11,39, минимальная - 3,4, максимальная - 36,1.

In [41]:
df.describe(include = 'all')

region municipality         year        birth  \
count                 4109         4109  4109.000000  4104.000000   
unique                  81          202          NaN          NaN   
top     Московская область      Барнаул          NaN          NaN   
freq                   365           27          NaN          NaN   
mean                   NaN          NaN  2007.474081    11.396564   
std                    NaN          NaN     7.686001     2.952200   
min                    NaN          NaN  1985.000000     3.400000   
25%                    NaN          NaN  2002.000000     9.300000   
50%                    NaN          NaN  2008.000000    10.900000   
75%                    NaN          NaN  2014.000000    13.000000   
max                    NaN          NaN  2019.000000    36.100000   

                 wage  
count     4098.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean     19619.858897  
std      18274.629869  
min          0.116000  
25%       4387.175000  
50%      15586.550000  
75%      30025.650000  
max     170460.000000

## Агрегирование
Если объяснять простыми словами, то агрегирование - это процесс приведения некого большого массива значений к некому сжатому по определенному параметру массиву значений. Например, среднее, медиана, количество, сумма.

In [42]:
df[['birth', 'wage']].agg(['mean', 'median', 'min', 'max'])

birth           wage
mean    11.396564   19619.858897
median  10.900000   15586.550000
min      3.400000       0.116000
max     36.100000  170460.000000

Мы убедились, что среднее значение по рождаемости - 11,39, средняя зарплата - 19619. Таким образом мы провели агрегирование.

## Groupby
Но средняя температура по больнице нам не интересна, мы хотим знать победителей в лицо. Допустим нам нужно посмотреть средние значения с группировкой по городам и субъектам РФ. Для этого закономерно используем метод .groupby(['region', 'municipality']).agg('mean').

In [43]:
df_groupby = df.groupby(['region', 'municipality']).agg('mean')
df_groupby.head(20)

year  \
region                       municipality                                   
Алтайский край               Барнаул                          2005.555556   
                             Бийск                            2009.500000   
                             Рубцовск                         2009.500000   
Амурская область             Благовещенск                     2005.555556   
Архангельская область        Архангельск                      2005.555556   
                             Северодвинск                     2009.500000   
Астраханская область         Город Астрахань                  2005.555556   
Белгородская область         Городской округ Белгород         2005.555556   
                             Старооскольский городской округ  2009.500000   
Брянская область             Город Брянск                     2005.555556   
Владимирская область         Город Владимир                   2005.555556   
                             Город Ковров                     2009.500000   
                             Округ Муром                      2009.500000   
Волгоградская область        Город Волгоград                  2005.555556   
                             Город Волжский                   2009.500000   
                             Город Камышин                    2009.500000   
Вологодская область          Город Вологда                    2005.555556   
                             Город Череповец                  2009.500000   
Воронежская область          город Воронеж                    2005.555556   
Еврейская автономная область Биробиджан                       2005.555556   

                                                                  birth  \
region                       municipality                                 
Алтайский край               Барнаул                          10.881481   
                             Бийск                            11.000000   
                             Рубцовск                          9.680000   
Амурская область             Благовещенск                     11.803704   
Архангельская область        Архангельск                      10.477778   
                             Северодвинск                     10.390000   
Астраханская область         Город Астрахань                  12.100000   
Белгородская область         Городской округ Белгород          9.977778   
                             Старооскольский городской округ  10.635000   
Брянская область             Город Брянск                      9.818519   
Владимирская область         Город Владимир                    9.566667   
                             Город Ковров                      9.290000   
                             Округ Муром                       9.490000   
Волгоградская область        Город Волгоград                   9.511111   
                             Город Волжский                    9.125000   
                             Город Камышин                     8.985000   
Вологодская область          Город Вологда                    11.759259   
                             Город Череповец                  11.540000   
Воронежская область          город Воронеж                     9.659259   
Еврейская автономная область Биробиджан                       12.366667   

                                                                      wage  
region                       municipality                                   
Алтайский край               Барнаул                          12700.674407  
                             Бийск                            13479.875000  
                             Рубцовск                         12774.935000  
Амурская область             Благовещенск                     17359.895556  
Архангельская область        Архангельск                      18911.154519  
                             Северодвинск                     27084.675000  
Астраханская область         Город Астрахань                  13896.580111  
Белгородская область         Г

Но мы на этом не успкаиваемся и хотим больше данных: среднее, медиану, минимум, максимум.

In [44]:
agg_func_math = {
    'birth': ['mean', 'median', 'min', 'max'],
    'wage': ['mean', 'median', 'min', 'max']
}

In [45]:
df.groupby(['region', 'municipality']).agg(agg_func_math).head(7)

birth                    \
                                            mean median  min   max   
region                municipality                                   
Алтайский край        Барнаул          10.881481  10.70  7.6  16.0   
                      Бийск            11.000000  10.70  8.7  13.5   
                      Рубцовск          9.680000   9.65  7.2  12.3   
Амурская область      Благовещенск     11.803704  11.80  9.3  16.1   
Архангельская область Архангельск      10.477778  10.70  7.9  15.4   
                      Северодвинск     10.390000  10.40  7.0  13.1   
Астраханская область  Город Астрахань  12.100000  12.30  8.9  14.5   

                                               wage                      \
                                               mean    median       min   
region                municipality                                        
Алтайский край        Барнаул          12700.674407   7696.00     0.213   
                      Бийск            13479.875000  12575.75  1386.000   
                      Рубцовск         12774.935000  11907.25  1192.000   
Амурская область      Благовещенск     17359.895556  11913.70     0.219   
Архангельская область Архангельск      18911.154519  13432.00     0.236   
                      Северодвинск     27084.675000  23156.80  2407.000   
Астраханская область  Город Астрахань  13896.580111   8448.80     0.170   

                                                
                                           max  
region                municipality              
Алтайский край        Барнаул          37855.1  
                      Бийск            30203.9  
                      Рубцовск         29941.9  
Амурская область      Благовещенск     50675.5  
Архангельская область Архангельск      54598.8  
                      Северодвинск     63519.0  
Астраханская область  Город Астрахань  39924.8

Посмотрим топ городов по зарплатам.

In [46]:
df.groupby(['region', 'municipality']).agg('mean').sort_values(by='wage', ascending=False).head(7)

year  \
region                                   municipality                 
Московская область                       Долгопрудный   2017.000000   
Ямало-Ненецкий автономный округ          Новый Уренгой  2010.277778   
Московская область                       Реутов         2017.000000   
                                         Домодедово     2014.500000   
Ханты-Мансийский автономный округ - Югра Сургут         2009.500000   
Красноярский край                        Норильск       2009.500000   
Ямало-Ненецкий автономный округ          Ноябрьск       2010.277778   

                                                            birth  \
region                                   municipality               
Московская область                       Долгопрудный   10.420000   
Ямало-Ненецкий автономный округ          Новый Уренгой  13.894444   
Московская область                       Реутов          9.360000   
                                         Домодедово     13.290000   
Ханты-Мансийский автономный округ - Югра Сургут         16.595000   
Красноярский край                        Норильск       12.700000   
Ямало-Ненецкий автономный округ          Ноябрьск       13.400000   

                                                                wage  
region                                   municipality                 
Московская область                       Долгопрудный   69366.200000  
Ямало-Ненецкий автономный округ          Новый Уренгой  64519.627778  
Московская область                       Реутов         53252.200000  
                                         Домодедово     52964.010000  
Ханты-Мансийский автономный округ - Югра Сургут         51231.465000  
Красноярский край                        Норильск       50389.215000  
Ямало-Ненецкий автономный округ          Ноябрьск       50348.477778

А что если посмотреть данные в разрезе по годам...?

## Pivot table

Ответим на этот вопрос чуть позже, а пока создадим сводную таблицу аналогичную той, что мы уже делали ранее.

In [47]:
df_pivot_table = df.pivot_table(index=['region', 'municipality'])
df_pivot_table.head(7)

birth          wage         year
region                municipality                                         
Алтайский край        Барнаул          10.881481  12700.674407  2005.555556
                      Бийск            11.000000  13479.875000  2009.500000
                      Рубцовск          9.680000  12774.935000  2009.500000
Амурская область      Благовещенск     11.803704  17359.895556  2005.555556
Архангельская область Архангельск      10.477778  18911.154519  2005.555556
                      Северодвинск     10.390000  27084.675000  2009.500000
Астраханская область  Город Астрахань  12.100000  13896.580111  2005.555556

Уже на этом этапе видно, что сводная таблица достаточно умная и сама агрегировала данные и посчитала средние значения.

Но если мы захотим расширить таблицу новыми значениями - медианой, то увидим, что конечный результат по структуре будет отличаться от того, что мы делали при использовании метода groupby. 

In [53]:
df_pivot_table = df1.pivot_table(index=['region', 'municipality'],
                                values=['year', 'pension'],
                                aggfunc=[np.mean, np.median])
df_pivot_table.head(7)

mean                 median  \
                                            pension         year   pension   
region                municipality                                           
Алтайский край        Барнаул           7682.250000  2005.555556   8027.55   
                      Бийск             7333.783333  2009.500000   7680.80   
                      Рубцовск          7186.183333  2009.500000   7521.10   
Амурская область      Благовещенск      8028.077778  2005.555556   8347.20   
Архангельская область Архангельск       9712.111111  2005.555556  10086.35   
                      Северодвинск     10540.638889  2009.500000  10861.70   
Астраханская область  Город Астрахань   7273.744444  2005.555556   7577.60   

                                               
                                         year  
region                municipality             
Алтайский край        Барнаул          2006.0  
                      Бийск            2009.5  
                      Рубцовск         2009.5  
Амурская область      Благовещенск     2006.0  
Архангельская область Архангельск      2006.0  
                      Северодвинск     2009.5  
Астраханская область  Город Астрахань  2006.0

Так мы плавно подошли к тем преимуществам, которые делает сводные таблицы швейцарским ножом для аналитиков. Если использую groupby мы "укрупняли" строки по городам, то сейчас мы можем "развернуть" столбец, например, 'year' (год) и посмотреть данные в разрезе по годам.   

In [55]:
df_pivot_table = df.pivot_table(index=['region', 'municipality'],
                                values=['wage'],
                                columns='year')
df_pivot_table.head(20)

wage         \
year                                                           1985   1990   
region                       municipality                                    
Алтайский край               Барнаул                          0.213  0.296   
                             Бийск                              NaN    NaN   
                             Рубцовск                           NaN    NaN   
Амурская область             Благовещенск                     0.219  0.361   
Архангельская область        Архангельск                      0.236  0.336   
                             Северодвинск                       NaN    NaN   
Астраханская область         Город Астрахань                  0.170  0.293   
Белгородская область         Городской округ Белгород         0.175  0.268   
                             Старооскольский городской округ    NaN    NaN   
Брянская область             Город Брянск                     0.187  0.278   
Владимирская область         Город Владимир                   0.188  0.291   
                             Город Ковров                       NaN    NaN   
                             Округ Муром                        NaN    NaN   
Волгоградская область        Город Волгоград                  0.181  0.270   
                             Город Волжский                     NaN    NaN   
                             Город Камышин                      NaN    NaN   
Вологодская область          Город Вологда                    0.184  0.289   
                             Город Череповец                    NaN    NaN   
Воронежская область          город Воронеж                    0.183  0.269   
Еврейская автономная область Биробиджан                       0.221  0.329   

                                                                          \
year                                                         1991   1995   
region                       municipality                                  
Алтайский край               Барнаул                          NaN  440.0   
                             Бийск                            NaN    NaN   
                             Рубцовск                         NaN    NaN   
Амурская область             Благовещенск                     NaN  613.7   
Архангельская область        Архангельск                      NaN  617.0   
                             Северодвинск                     NaN    NaN   
Астраханская область         Город Астрахань                  NaN  388.0   
Белгородская область         Городской округ Белгород         NaN  392.9   
                             Старооскольский городской округ  NaN    NaN   
Брянская область             Город Брянск                     NaN  362.6   
Владимирская область         Город Владимир                   NaN  399.0   
                             Город Ковров                     NaN    NaN   
                             Округ Муром                      NaN    NaN   
Волгоградская область        Город Волгоград                  NaN  444.8   
                             Город Волжский                   NaN    NaN   
                             Город Камышин                    NaN    NaN   
Вологодская область          Город Вологда                    NaN  541.0   
                             Город Череповец                  NaN    NaN   
Воронежская область          город Воронеж                    NaN  325.8   
Еврейская автономная область Биробиджан                       NaN  539.9   

                                                                             \
year                                                           1996    1997   
region                       municipality                                     
Алтайский край               Барнаул                          756.5   889.9   
                             Бийск                              NaN     NaN   
                             Рубцовск                           NaN     NaN   
Амурская область             Благовещенск    

И, конечно, данные в сводной таблице можно фильтровать. Создадим сводную таблицу и оставим в ней данные по городам, в которых рождаемость на 1000 человек превышает 12, зарплата выше 40.000 и отсортируем всё по убыванию рождаемости.

In [56]:
df2 = df.pivot_table(index=['region', 'municipality'])
df2.query("`birth`>12 and `wage`>40000").sort_values(by='birth', ascending=False)

birth  \
region                                   municipality               
Ханты-Мансийский автономный округ - Югра Сургут         16.595000   
Ямало-Ненецкий автономный округ          Салехард       15.896296   
Московская область                       Раменский      14.922222   
Ямало-Ненецкий автономный округ          Новый Уренгой  13.894444   
                                         Ноябрьск       13.400000   
Московская область                       Домодедово     13.290000   
Ханты-Мансийский автономный округ - Югра Нефтеюганск    13.250000   
Московская область                       Красногорский  13.141667   
Чукотский автономный округ               Анадырь        12.825926   
Красноярский край                        Норильск       12.700000   

                                                                wage  \
region                                   municipality                  
Ханты-Мансийский автономный округ - Югра Сургут         51231.465000   
Ямало-Ненецкий автономный округ          Салехард       40139.665926   
Московская область                       Раменский      42697.511111   
Ямало-Ненецкий автономный округ          Новый Уренгой  64519.627778   
                                         Ноябрьск       50348.477778   
Московская область                       Домодедово     52964.010000   
Ханты-Мансийский автономный округ - Югра Нефтеюганск    40221.240000   
Московская область                       Красногорский  47492.783333   
Чукотский автономный округ               Анадырь        45773.736815   
Красноярский край                        Норильск       50389.215000   

                                                               year  
region                                   municipality                
Ханты-Мансийский автономный округ - Югра Сургут         2009.500000  
Ямало-Ненецкий автономный округ          Салехард       2005.555556  
Московская область                       Раменский      2015.000000  
Ямало-Ненецкий автономный округ          Новый Уренгой  2010.277778  
                                         Ноябрьск       2010.277778  
Московская область                       Домодедово     2014.500000  
Ханты-Мансийский автономный округ - Югра Нефтеюганск    2009.500000  
Московская область                       Красногорский  2013.500000  
Чукотский автономный округ               Анадырь        2005.555556  
Красноярский край                        Норильск       2009.500000

## Pivot

Всё было замечательно, но в Pandas кроме pivot_table есть ещё просто pivot. Посмотрим что это за зверь и чем они отличаются. 

Создадим pivot: рождаемость в разрезе по регионам и годам.

In [57]:
df_pivot = df.pivot(index='region',
                    values='birth',
                    columns='year')
df_pivot.head(7)

ValueError: Index contains duplicate entries, cannot reshape

Мы получили ошибку "Index contains duplicate entries, cannot reshape". Что-то не так с индексами, попробуем создать pivot с индексами по городам, а не регионам.

In [ ]:
df_pivot = df.pivot(index='municipality',
                    values='birth',
                    columns='year')
df_pivot.head(7)

year                 1985  1990  1991  1995  1996  1997  1998  1999  2000  \
municipality                                                                
Абакан               16.8  14.4   NaN  10.4   9.7   9.2   9.3   9.0  10.1   
Альметьевский район   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   9.9   
Анадырь              15.8  13.5   NaN   9.5   7.8   9.1  12.0  11.0  10.5   
Ангарский район       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.0   
Арзамас               NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.5   
Артёмовский           NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.1   
Архангельск          15.4  12.1   NaN   8.6   8.8   8.3   8.9   7.9   8.3   

year                 2001  ...  2010  2011  2012  2013  2014  2015  2016  \
municipality               ...                                             
Абакан               10.0  ...  14.6  14.6  14.9  15.4  14.9  15.4  14.5   
Альметьевский район  10.7  ...  13.9  13.9  14.8  15.8  15.8  17.4  16.7   
Анадырь              11.4  ...  17.9  14.0  13.5  12.2  12.1  12.6  12.2   
Ангарский район       9.1  ...  11.9  12.6  13.3  13.2  13.2  14.3  13.6   
Арзамас               9.0  ...  11.0  10.9  12.1  11.5  11.2  13.1  12.9   
Артёмовский           8.6  ...  12.1  12.7  13.3  12.9  13.2  13.3  12.1   
Архангельск           9.4  ...  11.7  11.1  11.9  11.8  12.1  11.8  11.6   

year                 2017  2018  2019  
municipality                           
Абакан               13.5  12.2  11.5  
Альметьевский район  15.0  13.6  12.7  
Анадырь              11.7  11.3   8.9  
Ангарский район      11.7  10.8   9.4  
Арзамас              11.3  10.3   9.6  
Артёмовский          11.8  11.3  10.6  
Архангельск          10.0   9.1   8.1  

[7 rows x 28 columns]

Всё получилось. Как мы уже заметили всё дело в индексах. На самом деле всё в повторяющихся индексах. Как мы видили ранее в датафрейме есть позиции с несколькими городами в рамках одного субъекта, так получаются дублированные данные индексов с которыми не умеет работать pivot.

## Вывод:
Groupby, pivot и pivot_table удобные и привычные инструменты для работы с данными. Groupby позволяет кодом в одну строку получить агрегированные и сортированные данные, а pivot и pivot_table работать в более глубоком разрезе. Pivot_table предпочтителен, т.к. не ограничивает вас в уникальности значений в столбце индекса. И, конечно, все эти данные можно фильтровать под ваши запросы.
